xdp-document.c

```diff
@@ -791,3 +791,125 @@ xdp_document_load (GomRepository      *repository,
         load->pending = g_list_append (load->pending, g_object_ref (task));
     }
 }
+
+static void
+document_saved (GObject *source_object,
+                GAsyncResult *result,
+                gpointer data)
+{
+  GomResource *resource = GOM_RESOURCE (source_object);
+  g_autoptr (GTask) task = data;
+  GError *error = NULL;
+
+  if (!gom_resource_save_finish (resource, result, &error))
+    g_task_return_error (task, error);
+  else
+    g_task_return_pointer (task, g_object_ref (resource), g_object_unref);
+}
+
+typedef struct
+{
+  GTask *task;
+  gchar *uri;
+} UriData;
+
+static void
+find_one_for_uri_cb (GObject *source_object,
+                     GAsyncResult *res,
+                     gpointer user_data)
+{
+  GomRepository *repository = GOM_REPOSITORY (source_object);
+  g_autofree UriData *data = user_data;
+  g_autoptr (GTask) task = data->task;
+  g_autofree char *uri = data->uri;
+  g_autoptr (XdpDocument) doc = NULL;
+  g_autoptr (GError) error = NULL;
+
+  doc = (XdpDocument *)gom_repository_find_one_finish (repository, res, &error);
+
+  if (doc == NULL)
+    {
+      if (g_error_matches (error, GOM_ERROR, GOM_ERROR_REPOSITORY_EMPTY_RESULT))
+        {
+          doc = xdp_document_new (repository, uri);
+          g_hash_table_insert (documents, &doc->id, doc);
+          gom_resource_save_async (GOM_RESOURCE (doc), document_saved, g_object_ref (task));
+          g_clear_error (&error);
+        }
+      else
+        {
+          g_task_return_error (task, error);
+          error = NULL;
+        }
+    }
+  else
+    {
+      DocumentLoad *load;
+      GValue value = { 0, };
+      g_autoptr(GomFilter) filter = NULL;
+
+      load = g_new0 (DocumentLoad, 1);
+      load->id = doc->id;
+      load->pending = g_list_append (load->pending, g_object_ref (task));
+
+      g_hash_table_insert (loads, &load->id, load);
```

Comment:
This misses the case where id is in the hash already. Can't you use xdp_document_load() here which handles that case too.